clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -1981,4 +1992,145 @@ void handleFailure(Throwable throwable) {
+                // Handle server responses for particular groupId.
+                for (Map.Entry<String, KafkaFutureImpl<ConsumerGroupDescription>> entry : consumerGroupFutures.entrySet()) {
+                    final String groupId = entry.getKey();
+                    final KafkaFutureImpl<ConsumerGroupDescription> future = entry.getValue();
+                    final DescribeGroupsResponse.GroupMetadata groupMetadata = response.groups().get(groupId);
+                    final Errors groupError = groupMetadata.error();
+                    if (groupError != Errors.NONE) {
+                        future.completeExceptionally(groupError.exception());
+                        continue;
+                    }
+
+                    final String protocolType = groupMetadata.protocolType();
+                    if (protocolType.equals(ConsumerProtocol.PROTOCOL_TYPE) || protocolType.isEmpty()) {
+                        final List<DescribeGroupsResponse.GroupMember> members = groupMetadata.members();
+                        final List<MemberDescription> consumers = new ArrayList<>(members.size());
+
+                        for (DescribeGroupsResponse.GroupMember groupMember : members) {
+                            final PartitionAssignor.Assignment assignment =
+                                ConsumerProtocol.deserializeAssignment(
+                                    ByteBuffer.wrap(Utils.readBytes(groupMember.memberAssignment())));
+
+                            final MemberDescription memberDescription =
+                                new MemberDescription(
+                                    groupMember.clientId(),
+                                    groupMember.memberId(),
+                                    groupMember.clientHost(),
+                                    new MemberAssignment(assignment.partitions()));
+                            consumers.add(memberDescription);
+                        }
+                        final String protocol = groupMetadata.protocol();
+                        final ConsumerGroupDescription consumerGroupDescription =
+                            new ConsumerGroupDescription(groupId, protocolType.isEmpty(), consumers, protocol);
+                        future.complete(consumerGroupDescription);
+                    }
+                }
+            }
+
+            @Override
+            boolean handleUnsupportedVersionException(UnsupportedVersionException exception) {
+                return false;
+            }
+
+            @Override
+            void handleFailure(Throwable throwable) {
+                completeAllExceptionally(consumerGroupFutures.values(), throwable);
+            }
+        }, now);
+        return new DescribeConsumerGroupsResult(new HashMap<String, KafkaFuture<ConsumerGroupDescription>>(consumerGroupFutures));
+    }
+
+    @Override
+    public ListConsumerGroupsResult listConsumerGroups(ListConsumerGroupsOptions options) {
```

Comment:
Unfortunately, the consumer groups are not aggregated in the same way that topic metadata is. To get all the groups in the cluster, you have to send the ListGroups request to all nodes.

Reply:
I assume this is the same as `DescribeConsumerGroup` API. If changes are fine, I will updated this an `listConsumerGroupOffsets`

Reply:
The logic is different for ListGroups. We have to send a separate request to every broker in the cluster and then aggregate the results.

Reply:
@hachikuji trying to test this I find out that the result from this operation has to change (similar to how the Scala API is doing now) to respond with `Map<Node, KafkaFuture<List<ConsumerGroupListing>>>` instead, to list all responses from nodes. Does this make sense? I can update the KIP regarding this.

Reply:
I'd vote for doing the flattening internally than expose the node information in the results, as this is supposed to be internal implementation details that is better not leaking out. The Scala API returning `Map[Node, List[GroupOverview]]` was not a well designed one in my hind-sight.